pkg/compute/store/boltdb/store.go

```diff
@@ -0,0 +1,439 @@
+	execution.State = request.NewState
+	execution.Version += 1
+	execution.UpdateTime = time.Now()
+
+	// Having modified the execution, we're going to write it back over the top of the previous entry
+	// before appending a copy of the history to the history bucket
+
+	data, err := json.Marshal(execution)
+	if err != nil {
+		return err
+	}
+
+	executionsBucket := tx.Bucket([]byte(BucketExecutionsName))
+	err = executionsBucket.Put([]byte(execution.ID), data)
+	if err != nil {
+		return err
+	}
+
+	return s.appendHistory(tx, execution, previousState, request.Comment)
+}
+
+// Must be called where tx is a write transaction and adds a new history entry
+// to the bucket in history/execution-id/ with a key value derived from the
+// bucket sequence. Elsewhere we want to iterate through the history so to
+// make sure we get it back in creation order we use a three-digit sequence
+// number as bucket traversals happen in lexicographical order.
+func (s *Store) appendHistory(
+	tx *bolt.Tx,
+	updatedExecution store.Execution,
+	previousState store.ExecutionState, comment string) error {
+	historyBucket := tx.Bucket([]byte(BucketHistoryName))
+	executionHistoryBucket, err := historyBucket.CreateBucketIfNotExists([]byte(updatedExecution.ID))
+	if err != nil {
+		return err
+	}
+
+	historyEntry := store.ExecutionHistory{
+		ExecutionID:   updatedExecution.ID,
+		PreviousState: previousState,
+		NewState:      updatedExecution.State,
+		NewVersion:    updatedExecution.Version,
+		Comment:       comment,
+		Time:          updatedExecution.UpdateTime,
+	}
+	historyEntryData, err := json.Marshal(historyEntry)
+	if err != nil {
+		return err
+	}
+
+	// History entry keys within the execution bucket will be iterated over
+	// in lexographical order so we want to pad the numbers with leading 0s
+	// although this adds a hard limit of 999 history items for an execution
+	seqNum, err := executionHistoryBucket.NextSequence()
+	if err != nil {
+		return err
+	}
+	seq := fmt.Sprintf("%03d", seqNum)
+	return executionHistoryBucket.Put([]byte(seq), historyEntryData)
+}
+
+// DeleteExecution delete the execution, removes its history and removes it
+// from the job index (along with the job indexes bucket)
+func (s *Store) DeleteExecution(ctx context.Context, executionID string) error {
+	log.Ctx(ctx).Trace().
+		Str("ExecutionID", executionID).
+		Msg("boltdb.DeleteExecution")
+
+	return s.database.Update(func(tx *bolt.Tx) (err error) {
+		return s.deleteExecution(tx, executionID)
+	})
+}
+
+func (s *Store) deleteExecution(tx *bolt.Tx, executionID string) error {
+	var errorList error
+
+	execution, err := s.getExecution(tx, executionID)
+	if err != nil {
+		errorList = errors.Join(errorList, store.NewErrExecutionNotFound(executionID))
+	}
+
+	// Delete single execution entry
+	executionsBucket := tx.Bucket([]byte(BucketExecutionsName))
+	err = executionsBucket.Delete([]byte(executionID))
+	if err != nil {
+		errorList = errors.Join(errorList, err)
+	}
```

Comment:
If I understand correctly, an error returned within a transaction means the whole transaction will be rolled-back. Is there a value of appending errors instead of just returning the first encountered error?
Also we are not consistent as we fail early if the job index couldn't be deleted

Reply:
You're right about the transaction rollback, I think I was starting to experiment with the idea of returning errors separately and running with the partial changes, but that's definitely risky and will leave us in a bit of a mess. Will fix to return early as we do elsewhere.

Reply:
Fixed in 8532dc1d